strawberryfields/engine.py

```diff
@@ -221,6 +221,9 @@ def _run_program(self, prog, **kwargs):
             commands that were applied to the backend, the samples returned from the backend, and
             the samples as a dictionary with the measured modes as keys
         """
+        # `_run_program` is called in `BaseEngine._run` but should never be called
+        # from a `BaseEngine` object, in which case an error should be raised.
+        raise NotImplementedError("'run_program' method must be implemented on child class")
```

Comment:
Are we sure `abc.abstractmethod` won't raise something similar already?

Reply:
Oh, maybe. I basically added this to get rid of the type hint issue with calling `run_engine` from `BaseEngine._run`. Without it, my type-hinting engine assumes that `None` is being returned here, and thus complains. Perhaps I should just ignore it though. 🤔

Reply:
Oh, I see! In that case, I'm okay with either approach (although I guess it's better safe than sorry?).